Src/Web/Web.Shared.Net/AzureAppServiceRoleNameFromHostNameHeaderInitializer.cs

```diff
@@ -0,0 +1,173 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureAppServiceRoleNameFromHostNameHeaderInitializer" /> class.
+        /// </summary>
+        public AzureAppServiceRoleNameFromHostNameHeaderInitializer() : this(".azurewebsites.net")
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AzureAppServiceRoleNameFromHostNameHeaderInitializer" /> class.
+        /// </summary>
+        /// <param name="webAppSuffix">WebApp name suffix.</param>
+        public AzureAppServiceRoleNameFromHostNameHeaderInitializer(string webAppSuffix)
+        {
+            this.WebAppSuffix = webAppSuffix;
+            try
+            {
+                var result = Environment.GetEnvironmentVariable(WebAppHostNameEnvironmentVariable);
+
+                if (!string.IsNullOrEmpty(result) && result.EndsWith(this.WebAppSuffix, StringComparison.OrdinalIgnoreCase))
+                {
+                    result = result.Substring(0, result.Length - this.WebAppSuffix.Length);
+                }
+
+                this.roleName = result;
+            }
+            catch (Exception ex)
+            {
+                WebEventSource.Log.LogAzureAppServiceRoleNameFromHostNameHeaderInitializerWarning(ex.ToInvariantString());
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets suffix of website name. This must be changed when running in non public Azure region.
+        /// Default value (Public Cloud):  ".azurewebsites.net"
+        /// For US Gov Cloud: ".azurewebsites.us"
+        /// For Azure Germany: ".azurewebsites.de".
+        /// </summary>
+        public string WebAppSuffix { get; set; }
+
+        /// <summary>
+        /// Populates RoleName from the request telemetry associated with the http context.
+        /// If RoleName is empty on the request telemetry, it'll be updated as well so that other telemetry
+        /// belonging to the same requests gets it from request telemetry, without having to parse headers again.
+        /// </summary>
+        /// <remarks>
+        /// RoleName is attempted from every incoming request as opposed to doing this periodically. This is
+        /// done to ensure every request (and associated telemetry) gets the correct RoleName during slot swap.
+        /// </remarks>
+        /// <param name="telemetry">The telemetry item for which RoleName is to be set.</param>
+        public void Initialize(ITelemetry telemetry)
+        {
+            try
+            {
+                if (!string.IsNullOrEmpty(telemetry.Context.Cloud.RoleName))
+                {
+                    // RoleName is already populated.
+                    return;
+                }
+
+                string roleName = string.Empty;
+                var context = this.ResolvePlatformContext();
+
+                if (context != null)
+                {
+                    var request = this.GetRequestFromContext(context);
```

Comment:
HttpContext is not guaranteed to be available here (only for Request you can rely it's here - other telemtery items reported in async/after async calls will not get proper role name

Reply:
modified cached value to be updated so it always carries "last known rolename".